index.rst

```diff
@@ -338,4 +338,121 @@ smaller than a single alert message.
+the transit time and number of consumers.
+
+
+.. figure:: _static/transitTimeConsumers.png
+   :width: 55%
+   :align: center
+   :name: figure-17
+
+   Transit time of alerts to consumers against number of consumers.
+
+For the test with 10+1 consumers including postage stamp cutouts, the average transit time is very large.
+The consumers are unable to keep up with bursts of alerts every 39 seconds.  Below shows the transit time lag
+increasing for alert bursts for 300 visits.
+
+.. figure:: _static/slowConsumer.png
+   :width: 55%
+   :align: center
+   :name: figure-18
+
+   Length of time from Kafka to consumer receipt and processing in seconds (y axis) vs. visit number (x axis).
+
+Fortunately, even with the very slow consumer read time, the producer submission time does not appear to be
+affected.
+
+
+Thoughts and Recommendations
+----------------------------
+For most experiments and metrics measured, assuming that the size of alerts has been estimated accurately,
+there is a significant impact on Kafka compute utilization and
+timing when postage stamp cutouts are included in the alert packets.  Additional consideration should be
+given to alternatives for the distribution of stamps, if possible.  One alternative to including postage
+stamps in the stream, recommended by other users of Kafka for large messages, is to forego including the
+stamps in the stream and instead include a url to the location of the stamps from which the stamp can be
+accessed separately.  This would particularly alleviate the increased network traffic for each additional
+consumer reading a stream.  Another potentially possible alternative could be to separate the postage stamps
+into their own Kafka topic.  If not all primary consumers are required or expected to read postage stamp
+cutouts and pass them to downstream users, this could decrease the total network traffic, since not all
+consumers would pull from the larger stream of stamps.  A downstream filter could then unite alerts of interest
+with corresponding stamps.  The feasibility of this option for combining streams still needs to be further investigated.
+
+It is unclear why an increased number of consumers would significantly slow a single producer as to be unable
+to keep up with the 39 second separation between bursts.  With larger instance sizes, there should be no bottleneck
+in reading from disk and no bottleneck in the total network traffic.  Though the parallelization of compute per CCD, i.e.,
+an increased number of producers in parallel, should eliminate this problem, there is the possibility of
+performance improvement if this issue can be addressed.  Preliminary testing indicates that adjusting three
+parameters, ``buffer.memory``, ``batch.size``, and ``linger.ms``, together may help.  Using multiple partitions
+may also help.  All tests above use a single partition for a single topic.
```

Comment:
This at least partially speaks to my question above. I'd really like to understand better what's going on here — without really understanding the Kafka architecture, it seems bizarre that number of consumers should be so tightly coupled to producer performance. Even if we can address this by tweaking the configuration, it seems to be hinting at some weirdness in the underlying Kafka system which I'd really like to get to the bottom of.

Reply:
Yes, I think I need to run some tests adjusting those parameters and trying to have the number of partitions ~ the number of consumers with the number of producers ~ number of ccds.  I think that the partitions is key, since from further reading, most users have a large number of partitions and consumers that read in groups.